File: 6. Array/Easy/06. LeftRotateK.py
#Approach 1 (TempArray)
class Solution:
    def rotateArray(self, nums, k):
        n = len(nums)
        k = k % n
        temp = [0] * n
        for i in range(k, n):
            temp[i - k] = nums[i]
        for i in range(k):
            temp[n - k + i] = nums[i]
        for i in range(n):
            nums[i] = temp[i]


#Approach 2 (Reverse)
class Solution:
    def reverse(self, nums, start, end):
        while start < end:
            nums[start], nums[end] = nums[end], nums[start]
            start += 1
            end -= 1

    def rotateArray(self, nums, k):
        n = len(nums)
        k = k % n
        self.reverse(nums, 0, k - 1)
        self.reverse(nums, k, n - 1)
        self.reverse(nums, 0, n - 1)

File: 6. Array/Easy/06. LeftRotateK.cpp
//Approach 1 (TempArray)
class Solution {
public:
    void rotateArray(vector<int>& nums, int k) {
        int n = nums.size();
        k = k % n;
        vector<int> temp(n);
        for(int i = k; i < n; i++){
            temp[i - k] = nums[i];
        }
        for(int i = 0; i < k; i++){
            temp[n - k + i] = nums[i];
        }
        nums = temp;
    }
};


//Approach 2 (Reverse)
class Solution {
public:
    void reverse(vector<int>& nums, int start, int end){
        while(start < end){
            swap(nums[start], nums[end]);
            start++;
            end--;
        }
    }

    void rotateArray(vector<int>& nums, int k) {
        int n = nums.size();
        k = k % n;
        reverse(nums, 0, k - 1);
        reverse(nums, k, n - 1);
        reverse(nums, 0, n - 1);
    }
};